internal/apiclient/policyv2.go

```diff
@@ -0,0 +1,294 @@
+func NewPolicyV2APIClient(ctx *config.RunContext) (*PolicyV2APIClient, error) {
+	c := PolicyV2APIClient{
+		APIClient: APIClient{
+			endpoint: ctx.Config.TagPolicyAPIEndpoint, // use tag policy api endpoint for now.
+			apiKey:   ctx.Config.APIKey,
+			uuid:     ctx.UUID(),
+		},
+	}
+
+	prw, err := c.getPolicyResourceAllowlist()
+	if err != nil {
+		return nil, err
+	}
+	c.allowlist = prw
+
+	return &c, nil
+}
+
+// UploadPolicyData sends a filtered set of a project's resource information to Infracost Cloud and
+// potentially adds PolicySha and PastPolicySha to the project's metadata.
+func (c *PolicyV2APIClient) UploadPolicyData(project *schema.Project) error {
+	if project.Metadata == nil {
+		project.Metadata = &schema.ProjectMetadata{}
+	}
+
+	filteredResources := c.filterResources(project.PartialResources)
+	if len(filteredResources) > 0 {
+		sha, err := c.uploadProjectPolicyData(filteredResources)
+		if err != nil {
+			return fmt.Errorf("failed to upload filtered partial resources %w", err)
+		}
+		project.Metadata.PolicySha = sha
+	}
+
+	filteredPastResources := c.filterResources(project.PartialPastResources)
+	if len(filteredPastResources) > 0 {
+		sha, err := c.uploadProjectPolicyData(filteredPastResources)
+		if err != nil {
+			return fmt.Errorf("failed to upload filtered past partial resources %w", err)
+		}
+		project.Metadata.PastPolicySha = sha
+	}
+
+	return nil
+}
+
+func (c *PolicyV2APIClient) uploadProjectPolicyData(p2rs []policy2Resource) (string, error) {
+	q := `
+	mutation($policyResources: [PolicyResourceInput!]!) {
+		storePolicyResources(policyResources: $policyResources) {
+			sha
+		}
+	}
+	`
+
+	v := map[string]interface{}{
+		"policyResources": p2rs,
+	}
+
+	results, err := c.doQueries([]GraphQLQuery{{q, v}})
+	if err != nil {
+		return "", fmt.Errorf("query storePolicyResources failed  %w", err)
+	}
+
+	if len(results) > 0 {
+		if results[0].Get("errors").Exists() {
+			return "", fmt.Errorf("query storePolicyResources failed, received graphql error: %s", results[0].Get("errors").String())
+		}
+	}
+
+	data := results[0].Get("data")
+
+	var response struct {
+		AddFilteredResourceSet struct {
+			Sha string `json:"sha"`
+		} `json:"storePolicyResources"`
+	}
+
+	err = json.Unmarshal([]byte(data.Raw), &response)
+	if err != nil {
+		return "", fmt.Errorf("failed to unmarshal storePolicyResources %w", err)
+	}
+
+	return response.AddFilteredResourceSet.Sha, nil
+}
+
+// graphql doesn't really like map/dictionary parameters, so convert tags,
+// values, and refs to key/value arrays.
+
+type policy2Tag struct {
+	Key   string `json:"key"`
+	Value string `json:"value"`
+}
+
+type policy2Value struct {
+	Key   string          `json:"key"`
+	Value json.RawMessage `json:"value"`
+}
+
+type policy2Reference struct {
+	Key       string   `json:"key"`
+	Addresses []string `json:"addresses"`
+}
+
+type policy2Resource struct {
```

Comment:
could we change this to?

```go
type policy2Resource struct {
    *schema.ResourceData
    Values       []policy2Value           `json:"values"`
	References   []policy2Reference       `json:"references"`
	Metadata     policy2InfracostMetadata `json:"infracostMetadata"`
}
```

The embedding means we only need to set the information we need with the policies API and won't potentially mis a struct copy.

I'm not sure if this works for this usecase though and it might be more pain then it's worth - so feel free to ignore.

Reply:
I think it's better as is because the policy2Resource is coupled with the graphql api more that it is the ResourceData.  If we add a new field to resource data without changing the graphql, the api call will blow up.  since we're stubbing out the api that will only happen once it gets to prod.